Exception when cookie value contains an equals character

As [discussed on Gitter](https://gitter.im/fs2-http/Lobby?at=58d05591590d72c864eae8fe), cookie values can contain equals signs. This leads to a runtime exception.  For example, using the echo service in the fs2-http readme:

```
$ http :9090/echo 'Cookie:name=value=1'
HTTP/1.1 400 Bad Request
```

...results in:

```
[ERROR] java.lang.Throwable: Decoding of the request header failed: 
    Headers/Cookie: Expected tuple, got List(Right(name), Right(value), Right(1))
[ERROR] 	at spinoco.fs2.http.HttpRequest$.$anonfun$fromStream$2(HttpRequestOrResponse.scala:218)
[ERROR] 	at fs2.Stream.$anonfun$flatMap$1(Stream.scala:148)
...
```

A "real world" cookie that has this structure is Google Analytics `__utmz`.

--- a/http/src/main/scala/spinoco/protocol/http/codec/helper.scala
+++ b/http/src/main/scala/spinoco/protocol/http/codec/helper.scala
@@ -199,7 +199,7 @@ object helper {
 
   /**
     * Tuple A, B separated by `discriminator`. This will first search for discriminator, and if match
-    * will apply `A` codec nad `B` codec to remaining bits, if any.
+    * will apply `A` codec and `B` codec to remaining bits, if any.
     */
   def tuple[A, B](discriminator:ByteVector, acodec:Codec[A], bCodec:Codec[B]):Codec[(A, B)] = {
     def encode(in:(A, B)):Attempt[List[BitVector]] = {

--- a/http/src/main/scala/spinoco/protocol/http/header/value/HttpCookie.scala
+++ b/http/src/main/scala/spinoco/protocol/http/header/value/HttpCookie.scala
@@ -7,8 +7,10 @@ import scodec.codecs._
 import scodec.{Attempt, Codec, Err}
 import spinoco.protocol.common.util.attempt
 import spinoco.protocol.http.codec.helper._
+import spinoco.protocol.common.codec.takeWhile
 
 import scala.concurrent.duration._
+import scala.collection.immutable.HashSet
 
 /**
   * see http://tools.ietf.org/html/rfc6265
@@ -27,11 +29,23 @@ sealed case class HttpCookie(
 
 object HttpCookie {
 
+  object rfc2616 {
+    val controls: Seq[Byte] = (0 to 31).map(_.toByte) :+ 127.toByte
+    val separators = Seq[Byte]('=', ')', '(', '<', '>', '@', ',', ';', ':', '\\', '"', '/', '[', ']', '?', '{', '}', ' ', 9.toByte)
+    val nonTokens = HashSet.empty[Byte] ++ separators ++ controls
+
+    val token: Codec[String] = takeWhile(asciiString) { b => nonTokens.contains(b) == false }
+  }
+
+  val cookiePair: Codec[(String,String)] =
+    (rfc2616.token :: constant(_equal) :: trimmedAsciiString).as[(String,String)]
+  // https://tools.ietf.org/html/rfc6265#section-5.3
+
   val codec: Codec[HttpCookie] = {
     val kvTuple:Codec[(String,String)] = { choice (
         asciiConstant("secure").exmap ( _ => Attempt.successful("Secure" -> "") , { case (k,_)  => if (k == "Secure") Attempt.successful(()) else Attempt.failure(Err("Secure expected")) })
         , asciiConstant("httponly").exmap ( _ => Attempt.successful("HttpOnly" -> "") , { case (k,_)  => if (k == "HttpOnly") Attempt.successful(()) else Attempt.failure(Err("HttpOnly expected")) })
-        , tuple(_equal, trimmedAsciiString, trimmedAsciiString)
+        , cookiePair
       )
     }
 

--- a/http/src/test/scala/spinoco/protocol/http/codec/HeaderCodecSpec.scala
+++ b/http/src/test/scala/spinoco/protocol/http/codec/HeaderCodecSpec.scala
@@ -292,6 +292,9 @@ property("Accept-Ranges Header") = secure {
       , ("Cookie: SESSID=298zf09hf012fh2; httponly"
         , Cookie(HttpCookie("SESSID", "298zf09hf012fh2", None, None, None, None, false, httpOnly = true, Map.empty))
         , "Cookie: SESSID=298zf09hf012fh2; httponly")
+      , ("Cookie: __utmz=118721.140907.1.utmcsr=(direct)|utmccn=(direct)|utmcmd=(none); Currency=%22usd%22"
+        , Cookie(HttpCookie("__utmz", "118721.140907.1.utmcsr=(direct)|utmccn=(direct)|utmcmd=(none)", None, None, None, None, false, false, Map("Currency" -> "%22usd%22")))
+        , "Cookie: __utmz=118721.140907.1.utmcsr=(direct)|utmccn=(direct)|utmcmd=(none); Currency=%22usd%22")
     ))
   }
 

